### ISSUE TITLE
is the documentation about sampled_softmax_loss correct?

### ISSUE THREAD

**Issue by lpxz** [tensorflow/tensorflow#4904]
Hi, 
"tf.nn.sampled_softmax_loss(weights, biases, inputs, labels, num_sampled, num_classes, num_true=1, sampled_values=None, remove_accidental_hits=True, partition_strategy='mod', name='sampled_softmax_loss')"
the document says:
"At inference time, you can compute full softmax probabilities with the expression tf.nn.softmax(tf.matmul(inputs, tf.transpose(weights)) + biases)."

I am not sure if this describes the behavior correctly. First, the labels are not used in the softmax computation; Second, the weights matrix is merely a lookup table that maps each word to a vector (word2vec), the above matmul operation is not meaningful.

Here is my understanding of the code:
Given any word w, we have a vector encoding V1 from inputs, and then with label, we look up the weight matrix to find a vector encoding V2. Then we compute the softmax loss (cross-entropy loss) between V1 and V2. Finally, we sum up the loss for the sampled words.

**Comment by asimshankar** [comment]
@gouwsmeister : Could you comment on this?

**Comment by gouwsmeister** [comment]
Hi @lpxz,

`labels` are used both in the softmax computation and in the sampled softmax computation. Here is another more intuitive explanation that might be useful: 

Let each batch of the training data consist of (inputs, labels) examples. Here, inputs may be a (vector) of integer word ids (the "contexts"), and labels is a vector of batch_size target-class integer ids (the "target words"). 

We now assume your model has a network which maps the `input`s into fixed-length `dim`-dimensional vectors. This [batch_size, dim] matrix is the `inputs` argument to tf.nn.sampled_softmax_loss. For each of the num_classes classes, we furthermore have a (learned) `dim`-dimensional embedding vector. This [num_classes, dim] matrix is the `weights` argument (same with full softmax). Finally, we also pass in the vector of observed target-class ids described above, i.e. `labels`.

So far it's the same for both the normal softmax and the sampled softmax. The only difference is then that the sampled softmax only samples `num_sampled` so-called "negative" (i.e. not observed) classes for each observed class in `labels`, and then computes the cross-entropy loss between the observed (labels) and not observed (sampled) classes, whereas the full softmax uses all the actual classes for the "negative" part.

Regarding your second question: You are correct that `weights` is simply a lookup table (as defined above). The matmul computes the dot-product between each `dim`-dimensional row in `inputs`, and each `dim`-dimensional column ("output embedding") in `weights`. In each example, this dot product, together with the biases, represents the input "logit" to the softmax function (i.e. the z_j in the first equation in https://en.wikipedia.org/wiki/Softmax_function).

So I think the explanations are correct as is and I'll mark this closed for now. Please let me know if you have any other questions.

Stephan

**Comment by WalkerWalker** [comment]
@gouwsmeister 

Is tf.nn.sampled_softmax_loss() an implementation of negative sampling?

I always think they are different. Sampled softmax loss is a fast way to compute softmax (because the denominator of the softmax is huge) and negative sampling is ... emm ... conceptually "randomly select just a small number of “negative” words (let’s say 5) to update the weights for", i don't really know how to implement it. It's like a big black box to me. And just now some friend mentioned that we can use sampled softmax loss to implement negative sampling. I was shocked and did a little research about it.

It seems like they are from different paper and [What is candidate sampling](https://www.tensorflow.org/extras/candidate_sampling.pdf) describes that NCE is a generalized version of subsampled softmax and NEG is simplifying case of the NCE... this is just quite confusing. So I guess this is not the same but they are closely related and comparable?

I'm not sure if it's a good place to ask this. If not, please let me know and I will just delete it. thanks!

**Comment by Santosh-Gupta** [comment]
> hey @gouwsmeister
> 
> Thanks for your great explanation above. I have a further question. I'm not sure if it's a good place to ask this. If not, please let me know and I will just delete it. thanks!
> 
> **Is tf.nn.sampled_softmax_loss() an implementation of negative sampling?**
> 
> I always think they are different. Sampled softmax loss is a fast way to compute softmax (because the denominator of the softmax is huge) and negative sampling is ... emm ... conceptually "randomly select just a small number of “negative” words (let’s say 5) to update the weights for", i don't really know how to implement it. It's like a big black box to me. And just now some friend mentioned that we can use sampled softmax loss to implement negative sampling. I was shocked and did a little research about it.
> 
> It seems like they are from different paper and [What is candidate sampling](https://www.tensorflow.org/extras/candidate_sampling.pdf) describes that NCE is a generalized version of subsampled softmax and NEG is simplifying case of the NCE... this is just quite confusing. So I guess this is not the same but they are closely related and comparable?

Would love to hear more details as well

**Comment by ConanCui** [comment]
Hi @gouwsmeister 
I have noticed that the "num_true" in sampled_softmax_loss, which are the number of targets class, is fixed for each training instance. But in some situation, we have the different number for this, and I find here are an implementation for a dynamic "number true". Hope this can add in tensorflow.

TensorFlow op for the dynamic number of true classes per instance.
https://github.com/bxshi/dynamic_sampled_softmax_loss

**Comment by ucasiggcas** [comment]
how to test the results trained by the sampled_softmax_loss,
is there any metrics?
such as acc, precision, recall, auc ?
with the  sampled_softmax_loss function ,I could not get the metrics, 
could pls help me ?
if you can read Chinese , also see the [blog](https://iggcas.blog.csdn.net/article/details/106147091)

thx
